space-provider/kcp/kcp.go

```diff
@@ -193,69 +202,194 @@ func (k *KcpWatcher) ResultChan() <-chan clusterprovider.WatchEvent {
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	lc, ok := obj.(*kcpcorev1alpha1.LogicalCluster)
+	if !ok {
+		runtime.HandleError(errors.New("unexpected object type. expected LogicalCluster"))
+		return
+	}
+	path, ok := lc.Annotations[kcpcore.LogicalClusterPathAnnotationKey]
+	if ok && lc.Spec.Owner != nil {
+		c.logger.V(4).Info("queueing LogicalCluster", "key", key)
+		c.queue.Add(
+			queueItem{
+				key:   key,
+				path:  path,
+				owner: lc.Spec.Owner,
+			},
+		)
+	}
+
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("starting KCP provider controller")
+	defer c.logger.Info("shutting down KCP provider controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.ctx, c.runWorker, time.Second)
+	}
+
+	<-c.ctx.Done()
+}
+
+func (c *controller) runWorker(ctx context.Context) {
+	for c.processNextItem() {
+	}
+}
+
+func (c *controller) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	i, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	item := i.(queueItem)
+
+	// Done with this key, unblock other workers.
+	defer c.queue.Done(i)
+
+	if err := c.process(item); err != nil {
+		runtime.HandleError(err)
+		c.queue.AddRateLimited(i)
+		return true
+	}
+	c.queue.Forget(i)
+	return true
+}
+
+func (c *controller) process(item queueItem) error {
+	lc, exists, err := c.lcInformer.GetIndexer().GetByKey(item.key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		c.handleDelete(item)
+	} else {
+		c.handleAdd(lc, item)
+	}
+	return nil
+}
+
+func (c *controller) handleAdd(logicalCluster interface{}, item queueItem) {
+	lc, ok := logicalCluster.(*kcpcorev1alpha1.LogicalCluster)
+	if !ok {
+		runtime.HandleError(errors.New("unexpected object type. expected LogicalCluster"))
+		return
+	}
+	path, owner := item.path, item.owner
+	c.logger.V(2).Info("KCP workspace modify/add event", "ws", path, "status", lc.Status.Phase)
+	c.watcher.provider.lock.Lock()
+	defer c.watcher.provider.lock.Unlock()
+
+	_, ok = c.watcher.provider.workspaces[path]
+	if !ok && lc.Status.Phase == kcpcorev1alpha1.LogicalClusterPhaseReady {
+		spaceInfo, err := c.watcher.provider.Get(path)
+		if err != nil {
+			c.logger.Info("Failed to get space info")
+			return
+		}
+		c.logger.V(2).Info("New KCP workspace is ready", "ws", path)
+		// add ready WS to cache and send an event
+		c.watcher.provider.workspaces[path] = string(lc.Status.Phase)
+		c.watcher.ch <- clusterprovider.WatchEvent{
+			Type:      watch.Added,
+			Name:      owner.Name,
+			SpaceInfo: spaceInfo,
+		}
+	}
+	if ok && lc.Status.Phase != kcpcorev1alpha1.LogicalClusterPhaseReady {
+		c.logger.V(2).Info("KCP workspace is not ready")
+		if ok {
+			delete(c.watcher.provider.workspaces, path)
+			c.watcher.ch <- clusterprovider.WatchEvent{
+				Type: watch.Deleted,
```

Comment:
Can you clarify please why we are deleting the workspace and sending the watcher a delete event if the kcp logical cluster phase is not ready?

Reply:
This is the case when ready workspace became not ready for some reason. In this case we want to remove this workspace from the list of ready workspaces and send an event to the space-manager to set the space object phase to not ready.